File: 2017/day6/part2.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

int main(int argc, char *argv[])
{
	std::vector<int> blocks;
	int temp;
	while (std::cin >> temp)
		blocks.push_back(temp);
	int redis = 0, second_count = 0;
	bool finished = false, seen_state_before = false;
	std::set<std::vector<int> > comb_set, second_set;
	while (!finished) {
		auto max_elem = std::max_element(blocks.begin(), blocks.end());
		bool inner_finished = false;
		int max_elem_value = *max_elem;
		auto inner_iter = (max_elem + 1 == blocks.end() ? blocks.begin() : max_elem + 1);
		*max_elem = 0;
		while (!inner_finished) {
			*inner_iter += 1;
			max_elem_value--;
			inner_iter++;
			if (inner_iter == blocks.end()) inner_iter = blocks.begin();
			if (max_elem_value == 0) inner_finished = true;
		}
		if (comb_set.count(blocks) == 0 && !seen_state_before) {
			comb_set.insert(blocks);
			redis++;
		}
		else if (comb_set.count(blocks) == 1) {
			seen_state_before = true;
			if (second_set.count(blocks) == 0) {
				second_set.insert(blocks);
				second_count++;
			}
			else if (second_set.count(blocks) == 1) {
				finished = true;
			}
		}
	}
	std::cout << second_count << '\n';
	return 0;
}

// int sz = blocks.size() - 1
// int each_gets = (*max_elem) / sz;

File: 2017/day6/part2.swift
import Foundation

func part2(_ input: String) -> Int {
    var inputSplit = input.components(separatedBy: CharacterSet.whitespaces).compactMap { Int(String($0)) }
    var ret = 0
    var finished = false, seenStateBefore = false
    var combinationSet = Set<[Int]>(), secondSeen = Set<[Int]>()
    while !finished {
        var maxElem = inputSplit.max()!
        var innerFinished = false
        let maxElemIdx = inputSplit.firstIndex(of: maxElem)!
        var innerIter = maxElemIdx + 1 == inputSplit.count ? 0 : maxElemIdx + 1
        inputSplit[maxElemIdx] = 0
        while !innerFinished {
            inputSplit[innerIter] += 1
            maxElem -= 1
            innerIter += 1
            if innerIter == inputSplit.count {
                innerIter = 0
            }
            if maxElem == 0 {
                innerFinished = true
            }
        }
        if !combinationSet.contains(inputSplit) && !seenStateBefore {
            combinationSet.insert(inputSplit)
        } else if combinationSet.contains(inputSplit) {
            seenStateBefore = true
            if !secondSeen.contains(inputSplit) {
                secondSeen.insert(inputSplit)
                ret += 1
            } else {
                finished = true
            }
        }
    }
    return ret
}
